gobblin-data-management/src/main/java/gobblin/data/management/copy/CopyableFile.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (C) 2014-2015 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.data.management.copy;
+
+import gobblin.configuration.ConfigurationKeys;
+import gobblin.data.management.partition.File;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.DataInput;
+import java.io.DataInputStream;
+import java.io.DataOutput;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.util.List;
+import java.util.Properties;
+
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+import org.apache.commons.codec.DecoderException;
+import org.apache.commons.codec.binary.Hex;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.io.Writable;
+
+import com.google.common.collect.Lists;
+
+
+/**
+ * Abstraction for a file to copy from {@link #origin} to {@link #destination}.
+ */
+@Getter
+@Setter
+@AllArgsConstructor
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+@EqualsAndHashCode
+public class CopyableFile implements File, Writable {
+
+  public static final String SERIALIZED_COPYABLE_FILE = "gobblin.copy.serialized.copyable.file";
+
+  /** {@link FileStatus} of the existing origin file. */
+  private FileStatus origin;
+  /** Destination {@link Path} of the file. */
+  private Path destination;
+  /** {@link Path} to the file relative to the final publish directory {@link ConfigurationKeys#DATA_PUBLISHER_FINAL_DIR} */
+  private Path relativeDestination;
+  /** Desired {@link OwnerAndPermission} of the destination path. */
+  private OwnerAndPermission destinationOwnerAndPermission;
+  /**
+   * Desired {@link OwnerAndPermission} of the ancestor directories of the destination path. The list is ordered from
+   * deepest to highest directory.
+   *
+   * <p>
+   * For example, if {@link #destination} is /a/b/c/file, then the first element of this list is the desired owner and
+   * permission for directory /a/b/c, the second is the desired owner and permission for directory /a/b, and so on.
+   * </p>
+   *
+   * <p>
+   * If there are fewer elements in the list than ancestor directories in {@link #destination}, it is understood that
+   * extra directories are allowed to have any owner and permission.
+   * </p>
+   */
+  private List<OwnerAndPermission> ancestorsOwnerAndPermission;
```

Comment:
if we are copying a directory, we probably want to replicate the attributes of each subdirectory. Since we only copy files, for each file we need to store the attributes of its parent directories at least up to the directory that is being copied.